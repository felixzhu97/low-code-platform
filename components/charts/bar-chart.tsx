"use client"

import { useMemo } from "react"
import {
  BarChart as RechartsBarChart,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts"

interface BarChartProps {
  data: any[]
  xField: string
  yField: string
  seriesField?: string
  title?: string
  height?: number
  width?: number
  colors?: string[]
  showGrid?: boolean
  showLegend?: boolean
  showTooltip?: boolean
  isStacked?: boolean
}

export function BarChart({
  data,
  xField,
  yField,
  seriesField,
  title,
  height = 300,
  width = 500,
  colors = ["#0088FE", "#00C49F", "#FFBB28", "#FF8042", "#8884D8"],
  showGrid = true,
  showLegend = true,
  showTooltip = true,
  isStacked = false,
}: BarChartProps) {
  // 处理数据，确保格式正确
  const processedData = useMemo(() => {
    if (!data || !Array.isArray(data)) {
      return []
    }

    // 如果没有seriesField，直接返回原始数据
    if (!seriesField) {
      return data
    }

    // 如果有seriesField，需要处理数据格式
    const uniqueXValues = Array.from(new Set(data.map((item) => item[xField])))
    const uniqueSeriesValues = Array.from(new Set(data.map((item) => item[seriesField])))

    return uniqueXValues.map((xValue) => {
      const result: any = { [xField]: xValue }

      uniqueSeriesValues.forEach((seriesValue) => {
        const matchingItem = data.find((item) => item[xField] === xValue && item[seriesField] === seriesValue)
        result[seriesValue] = matchingItem ? matchingItem[yField] : 0
      })

      return result
    })
  }, [data, xField, yField, seriesField])

  // 生成图表的bars
  const renderBars = useMemo(() => {
    if (!seriesField) {
      return <Bar dataKey={yField} fill={colors[0]} />
    }

    const uniqueSeriesValues = Array.from(new Set(data.map((item) => item[seriesField])))

    return uniqueSeriesValues.map((seriesValue, index) => (
      <Bar
        key={seriesValue as string}
        dataKey={seriesValue as string}
        fill={colors[index % colors.length]}
        stackId={isStacked ? "stack" : undefined}
      />
    ))
  }, [data, yField, seriesField, colors, isStacked])

  if (!data || data.length === 0) {
    return (
      <div
        className="flex items-center justify-center rounded-md border border-dashed p-4 text-center text-muted-foreground"
        style={{ width: width || "100%", height: height || 300 }}
      >
        暂无数据
      </div>
    )
  }

  return (
    <div className="w-full">
      {title && <h3 className="mb-2 text-lg font-medium">{title}</h3>}
      <div style={{ width: width || "100%", height: height || 300 }}>
        <ResponsiveContainer width="100%" height="100%">
          <RechartsBarChart data={processedData} margin={{ top: 20, right: 30, left: 20, bottom: 5 }}>
            {showGrid && <CartesianGrid strokeDasharray="3 3" />}
            <XAxis dataKey={xField} />
            <YAxis />
            {showTooltip && <Tooltip />}
            {showLegend && <Legend />}
            {renderBars}
          </RechartsBarChart>
        </ResponsiveContainer>
      </div>
    </div>
  )
}
