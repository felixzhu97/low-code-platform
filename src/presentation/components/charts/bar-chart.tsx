"use client";

import React from "react";
import {
  BarChart as RechartsBarChart,
  Bar,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";

interface BarChartProps {
  data?: Array<Record<string, any>>;
  width?: number;
  height?: number;
  xAxisKey?: string;
  dataKeys?: string[];
  colors?: string[];
  title?: string;
}

// 默认数据
const defaultData = [
  { name: "1月", sales: 8500, revenue: 6200 },
  { name: "2月", sales: 9200, revenue: 7300 },
  { name: "3月", sales: 10800, revenue: 8600 },
  { name: "4月", sales: 9800, revenue: 7900 },
  { name: "5月", sales: 10200, revenue: 8100 },
  { name: "6月", sales: 9500, revenue: 7600 },
];

export function BarChart({
  data,
  width = 400,
  height = 300,
  xAxisKey = "name",
  title,
}: BarChartProps) {
  const chartData = data && data.length > 0 ? data : defaultData;

  return (
    <div className="w-full h-full">
      {title && <h3 className="text-lg font-semibold mb-2">{title}</h3>}
      <ResponsiveContainer width="100%" height={height}>
        <RechartsBarChart
          data={chartData}
          margin={{
            top: 20,
            right: 30,
            left: 20,
            bottom: 5,
          }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis dataKey={xAxisKey} />
          <YAxis />
          <Tooltip />
          <Legend />
          <Bar dataKey="sales" fill="#8884d8" name="销售额" />
          <Bar dataKey="revenue" fill="#82ca9d" name="收入" />
        </RechartsBarChart>
      </ResponsiveContainer>
    </div>
  );
}

export default BarChart;
