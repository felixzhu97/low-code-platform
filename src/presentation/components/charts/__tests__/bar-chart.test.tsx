import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { BarChart } from "../bar-chart";

// Mock recharts
vi.mock("recharts", () => ({
  BarChart: ({ children, data, ...props }: any) => (
    <div data-testid="recharts-bar-chart" data-props={JSON.stringify(props)}>
      {children}
      <div data-testid="chart-data">{JSON.stringify(data)}</div>
    </div>
  ),
  Bar: ({ dataKey, fill, name, ...props }: any) => (
    <div
      data-testid="recharts-bar"
      data-datakey={dataKey}
      data-fill={fill}
      data-name={name}
      {...props}
    />
  ),
  XAxis: ({ dataKey, ...props }: any) => (
    <div data-testid="recharts-x-axis" data-datakey={dataKey} {...props} />
  ),
  YAxis: (props: any) => <div data-testid="recharts-y-axis" {...props} />,
  CartesianGrid: (props: any) => (
    <div data-testid="recharts-cartesian-grid" {...props} />
  ),
  Tooltip: (props: any) => <div data-testid="recharts-tooltip" {...props} />,
  Legend: (props: any) => <div data-testid="recharts-legend" {...props} />,
  ResponsiveContainer: ({ children, width, height, ...props }: any) => (
    <div
      data-testid="recharts-responsive-container"
      data-width={width}
      data-height={height}
      {...props}
    >
      {children}
    </div>
  ),
}));

describe("BarChart", () => {
  const sampleData = [
    { name: "1月", sales: 8500, revenue: 6200 },
    { name: "2月", sales: 9200, revenue: 7300 },
    { name: "3月", sales: 10800, revenue: 8600 },
  ];

  describe("Rendering", () => {
    it("renders with default props", () => {
      render(<BarChart />);

      expect(
        screen.getByTestId("recharts-responsive-container")
      ).toBeInTheDocument();
      expect(screen.getByTestId("recharts-bar-chart")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-x-axis")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-y-axis")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-cartesian-grid")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-tooltip")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-legend")).toBeInTheDocument();
    });

    it("renders with custom data", () => {
      render(<BarChart data={sampleData} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement).toBeInTheDocument();
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("8500");
    });

    it("renders with title", () => {
      const title = "Sales Chart";
      render(<BarChart title={title} />);

      expect(screen.getByText(title)).toBeInTheDocument();
    });

    it("renders with custom dimensions", () => {
      render(<BarChart width={600} height={400} />);

      const container = screen.getByTestId("recharts-responsive-container");
      expect(container).toHaveAttribute("data-width", "100%");
      expect(container).toHaveAttribute("data-height", "400");
    });

    it("renders with custom xAxisKey", () => {
      render(<BarChart data={sampleData} xAxisKey="date" />);

      const xAxis = screen.getByTestId("recharts-x-axis");
      expect(xAxis).toHaveAttribute("data-datakey", "date");
    });

    it("renders sales and revenue bars", () => {
      render(<BarChart data={sampleData} />);

      const bars = screen.getAllByTestId("recharts-bar");
      expect(bars).toHaveLength(2);

      expect(bars[0]).toHaveAttribute("data-datakey", "sales");
      expect(bars[0]).toHaveAttribute("data-fill", "#8884d8");
      expect(bars[0]).toHaveAttribute("data-name", "销售额");

      expect(bars[1]).toHaveAttribute("data-datakey", "revenue");
      expect(bars[1]).toHaveAttribute("data-fill", "#82ca9d");
      expect(bars[1]).toHaveAttribute("data-name", "收入");
    });
  });

  describe("Data Handling", () => {
    it("uses default data when no data provided", () => {
      render(<BarChart />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("6月");
    });

    it("uses default data when empty array provided", () => {
      render(<BarChart data={[]} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("6月");
    });

    it("uses provided data when valid data given", () => {
      render(<BarChart data={sampleData} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("3月");
      expect(chartDataElement.textContent).not.toContain("6月");
    });
  });

  describe("Props Configuration", () => {
    it("applies correct chart margins", () => {
      render(<BarChart data={sampleData} />);

      const chartElement = screen.getByTestId("recharts-bar-chart");
      const props = JSON.parse(chartElement.getAttribute("data-props") || "{}");

      expect(props.margin).toEqual({
        top: 20,
        right: 30,
        left: 20,
        bottom: 5,
      });
    });

    it("sets correct default dimensions", () => {
      render(<BarChart />);

      const container = screen.getByTestId("recharts-responsive-container");
      expect(container).toHaveAttribute("data-width", "100%");
      expect(container).toHaveAttribute("data-height", "300");
    });
  });

  describe("Accessibility", () => {
    it("has proper heading structure when title is provided", () => {
      const title = "Monthly Sales Report";
      render(<BarChart title={title} />);

      const heading = screen.getByRole("heading", { level: 3 });
      expect(heading).toHaveTextContent(title);
    });

    it("does not render heading when no title provided", () => {
      render(<BarChart />);

      const heading = screen.queryByRole("heading");
      expect(heading).not.toBeInTheDocument();
    });
  });

  describe("Snapshot Tests", () => {
    it("matches snapshot with default props", () => {
      const { container } = render(<BarChart />);
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot with custom data and title", () => {
      const { container } = render(
        <BarChart
          data={sampleData}
          title="Custom Chart"
          width={500}
          height={350}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot with all props", () => {
      const { container } = render(
        <BarChart
          data={sampleData}
          width={800}
          height={400}
          xAxisKey="name"
          title="Complete Bar Chart"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
