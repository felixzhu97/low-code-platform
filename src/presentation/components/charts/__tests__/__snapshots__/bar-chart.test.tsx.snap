// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarChart > Snapshot Tests > matches snapshot with all props 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <h3
      class="text-lg font-semibold mb-2"
    >
      Complete Bar Chart
    </h3>
    <div
      data-height="400"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-bar-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="sales"
          data-fill="#8884d8"
          data-name="销售额"
          data-testid="recharts-bar"
        />
        <div
          data-datakey="revenue"
          data-fill="#82ca9d"
          data-name="收入"
          data-testid="recharts-bar"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","sales":8500,"revenue":6200},{"name":"2月","sales":9200,"revenue":7300},{"name":"3月","sales":10800,"revenue":8600}]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BarChart > Snapshot Tests > matches snapshot with custom data and title 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <h3
      class="text-lg font-semibold mb-2"
    >
      Custom Chart
    </h3>
    <div
      data-height="350"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-bar-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="sales"
          data-fill="#8884d8"
          data-name="销售额"
          data-testid="recharts-bar"
        />
        <div
          data-datakey="revenue"
          data-fill="#82ca9d"
          data-name="收入"
          data-testid="recharts-bar"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","sales":8500,"revenue":6200},{"name":"2月","sales":9200,"revenue":7300},{"name":"3月","sales":10800,"revenue":8600}]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BarChart > Snapshot Tests > matches snapshot with default props 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <div
      data-height="300"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-bar-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="sales"
          data-fill="#8884d8"
          data-name="销售额"
          data-testid="recharts-bar"
        />
        <div
          data-datakey="revenue"
          data-fill="#82ca9d"
          data-name="收入"
          data-testid="recharts-bar"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","sales":8500,"revenue":6200},{"name":"2月","sales":9200,"revenue":7300},{"name":"3月","sales":10800,"revenue":8600},{"name":"4月","sales":9800,"revenue":7900},{"name":"5月","sales":10200,"revenue":8100},{"name":"6月","sales":9500,"revenue":7600}]
        </div>
      </div>
    </div>
  </div>
</div>
`;
